learn/chapter1/2-hello-triangle.md

```diff
@@ -0,0 +1,463 @@
+
+```cs
+float[] vertices = {
+    -0.5f, -0.5f, 0.0f, //Bottom-left vertex
+     0.5f, -0.5f, 0.0f, //Bottom-right vertex
+     0.0f,  0.5f, 0.0f  //Top vertex
+};
+```
+
+Because OpenGL works in 3D space we render a 2D triangle with each vertex having a z coordinate of 0.0. This way the depth of the triangle remains the same making it look like it's 2D.
+
+## Normalized Device Coordinates (NDC)
+
+Once your vertex coordinates have been processed in the vertex shader, they should be in normalized device coordinates which is a small space where the x, y and z values vary from -1.0 to 1.0. Any coordinates that fall outside this range will be discarded/clipped and won't be visible on your screen. Below you can see the triangle we specified within normalized device coordinates (ignoring the z axis):
+
+2D Normalized Device Coordinates as shown in a graph
+
+![NDC Graph](img/2-ndc.png)
+
+Unlike usual screen coordinates the positive y-axis points in the up-direction and the (0,0) coordinates are at the center of the graph, instead of top-left. Eventually you want all the (transformed) coordinates to end up in this coordinate space, otherwise they won't be visible.
+
+Your NDC coordinates will then be transformed to screen-space coordinates via the viewport transform using the data you provided with `GL.Viewport`. The resulting screen-space coordinates are then transformed to fragments as inputs to your fragment shader.
+
+## Buffers
+
+With the vertex data defined we'd like to send it as input to the first process of the graphics pipeline: the vertex shader. This is done by creating memory on the GPU where we store the vertex data, configure how OpenGL should interpret the memory and specify how to send the data to the graphics card. The vertex shader then processes as much vertices as we tell it to from its memory.
+
+We manage this memory via so called vertex buffer objects (VBO) that can store a large number of vertices in the GPU's memory. The advantage of using those buffer objects is that we can send large batches of data all at once to the graphics card without having to send data a vertex a time. Sending data to the graphics card from the CPU is relatively slow, so wherever we can, we try to send as much data as possible at once. Once the data is in the graphics card's memory the vertex shader has almost instant access to the vertices making it extremely fast.
+
+A vertex buffer object is our first occurrence of an OpenGL object as we've discussed in the OpenGL tutorial. Just like any object in OpenGL this buffer has a unique ID corresponding to that buffer, so we can generate one with a buffer ID using the `GL.GenBuffers` function.
+
+Add an int to your Game class to store the handle
+
+`int VertexBufferObject;`
```

Comment:
These one line snippets should be blocks given that they have their own line.

Author: @devvoid

Reply:
Did this for all of them I could find in #29